File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/network/base/PersistentClient.java
package net.namekdev.entity_tracker.network.base;


/**
 * Client that automatically reconnects.
 *
 * @author Namek
 */
public class PersistentClient extends Client {
    private volatile boolean isReconnectEnabled;

    /**
     * Delay between two reconnects, specified in milliseconds.
     */
    public int reconnectDelay = 1000;


    public PersistentClient(RawConnectionCommunicator connectionListener) {
        super.setConnectionListener(connectionListener);
    }

    @Override
    public Client connect(String serverName, int serverPort) {
        return connect(serverName, serverPort, false);
    }

    public Client connect(final String serverName, final int serverPort, final boolean manualUpdate) {
        isReconnectEnabled = true;

        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                tryConnect();

                while (isReconnectEnabled) {
                    if (!isConnected()) {
                        tryConnect();
                    }

                    try {
                        Thread.sleep(reconnectDelay);
                    } catch (InterruptedException e) {
                    }
                }
            }

            private void tryConnect() {
                try {
                    PersistentClient.super.connect(serverName, serverPort);

                    if (!manualUpdate)
                        PersistentClient.super.startThread();
                } catch (Exception ex) {
                }
            }
        });

        thread.start();

        return this;
    }

    @Override
    public void stop() {
        super.stop();
        isReconnectEnabled = false;
    }
}

File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/network/base/PersistentClient.kt
package net.namekdev.entity_tracker.network.base


/**
 * Client that automatically reconnects.

 * @author Namek
 */
class PersistentClient(connectionListener: RawConnectionCommunicator) : Client() {
    @Volatile private var isReconnectEnabled: Boolean = false

    /**
     * Delay between two reconnects, specified in milliseconds.
     */
    var reconnectDelay = 1000


    init {
        super.connectionListener = connectionListener
    }

    override fun connect(serverName: String, serverPort: Int): Client {
        return connect(serverName, serverPort, false)
    }

    fun connect(serverName: String, serverPort: Int, manualUpdate: Boolean): Client {
        isReconnectEnabled = true

        val thread = Thread(object : Runnable {
            override fun run() {
                tryConnect()

                while (isReconnectEnabled) {
                    if (!isConnected) {
                        tryConnect()
                    }

                    try {
                        Thread.sleep(reconnectDelay.toLong())
                    }
                    catch (e: InterruptedException) {
                    }

                }
            }

            private fun tryConnect() {
                try {
                    super@PersistentClient.connect(serverName, serverPort)

                    if (!manualUpdate)
                        super@PersistentClient.startThread()
                }
                catch (ex: Exception) {
                }

            }
        })

        thread.start()

        return this
    }

    override fun stop() {
        super.stop()
        isReconnectEnabled = false
    }
}
